### DISCUSSION TITLE
Current Django locale not honored

### DISCUSSION THREAD

**Discussion by arneatec** [codingjoe/django-select2#162] [Q&A] [answered]
**Issue**
django-select2 does not use the current active translation of Django.

Select2Mixin is only initialized on (development) server start this way:
```python
def __init__(self, *args, **kwargs):
    super().__init__(*args, **kwargs)
    self.i18n_name = SELECT2_TRANSLATIONS.get(get_language())
```
`get_language()` will make the best effort and return a language (for example, from the default `settings.LANGUAGE_CODE = "en"`).

But the current language might change on a request basis, for example due to `django.middleware.locale.LocaleMiddleware` with a browser with `Accept-Language: bg,en-US;q=0.9,en;q=0.8` header.

`LocaleMiddleware` will activate the 'bg' language on receiving the request and Django will translate the form to Bulgarian, but the self.i18n_name will remain 'en' for django-select2. 

The `Select2Mixin.media()` method will thus return the 'en' language for the JS translation file. 
```python
@property
def media(self):
    ... code removed for brevity ...
    i18n_file = []
    
    if self.i18n_name in settings.SELECT2_I18N_AVAILABLE_LANGUAGES:
        i18n_file = [f"{settings.SELECT2_I18N_PATH}/{self.i18n_name}.js"]
```
This will prevent the select2-based controls from translating correctly. No amount of effort (e.g. setting _data-language_ or _lang_ in the _attrs_ for a control) can fix this, as you would also be required to resolve the JS file name externally and swap it with the _incorrect_ one being supplied by the _media_ method.

**Expected behavior**
The django-select2 forms should be translated transparently as the rest of the form is translated.

django-select2 should check the current active translation before building the _media_ information and also before _build_attrs_

**Comment by codingjoe** [comment]
Hi @arneatec thank you for reaching out. I sincerely appreciate your detailed report. Would you be so kind as to explain how exactly you are experiencing this problem? Your deduction is interesting, but Django handles languages actually slightly different from what you describe. Django uses thread locals and the locale middleware will set the variable.

As it happens, I implemented the same language behavior for select2 in Django itself for autocomplete fields, so I am fairly confident about the implementation. But I'm known to make mistakes 😜 

However, let's try to troubleshoot this together. Some issues I could think of would be that the language isn't properly set OR some kind of state leakage. It would help you have some browser logs, that include the request headers and cookies.

I hope we can figure this out!

Best,
Joe

**Reply by arneatec** [reply]
Hi @codingjoe ,

I'll create a simplified test project in repository to make sure this is reproducible and get back to you when ready, Thank you for your prompt reply.

Regards,
Asen

**Comment by arneatec** [comment, marked as answer]
Hi @codingjoe 

Simplified code is here https://github.com/arneatec/django-select2-locale.git

I've used Bulgarian (bg) as test language. Model field names are translated to Bulgarian (with the English captions put in parentheses after the translation for your convenience).

I'm using pycharm with runserver. Requirements file shows there are no other fancy dependencies that might be interfering, I've tried to make this as generic as possible.

Issue is reproducible in the example code. 

**Steps to reproduce**:

1. Using Chrome, add Bulgarian as a preferred language and put it on top of the list.
2. Start the sample project. Visit http://127.0.0.1:8000/blog/post_add/
3. Page heading should read "Създай нова книга (Create a new Book)". 
4. If it is not and is English-only, Bulgarian is not the preferred language (make sure the request header is  Accept-Language: bg, etc.) or you may have another issue with languages. Fix this.
5. If it is, django.middleware.locale.LocaleMiddleware is correctly processing the Accept-Language header and setting the locale.
6. Click the **Тема (Topic):** dropdown, which is the select2 widget.
7. The caption shows 'Please enter 2 or more characters', which is NOT CORRECT as it should show the Bulgarian translation.
8. Go to the settings.py file and amend LANGUAGE_CODE as shown
  `LANGUAGE_CODE = 'bg' `
9. Restart server /rerun project. Visit http://127.0.0.1:8000/blog/post_add/
10. Click the **Тема (Topic):** dropdown, which is the select2 widget.
7. The caption shows 'Моля въведете още 2 символа', which is the CORRECT Bulgarian translation.
8. Inspect the html source. Note the included 
`<script src="https://cdnjs.cloudflare.com/ajax/libs/select2/4.0.12/js/i18n/bg.js"></script>`


**Workaround/temporary fix**:
Don't use self.i18n_name Select2Mixin member variable, use SELECT2_TRANSLATIONS.get(get_language()) directly.

Please let me know if I can be of further help.

Regards,
Asen

**Reply by codingjoe** [reply]
Thanks @arneatec, that really helped! I found the error, and I am working on a fix in #164. I'd really appreciate your feedback there.
